### DISCUSSION TITLE
trpc with existing react-query

### DISCUSSION THREAD

**Discussion by yikkwongwong** [trpc/trpc#3999] [Q&A] [answered]
### Area of Improvement

Not sure whether it's an error solely from my repo.

In order to integrate with existing react-query, I needed to make a new `QueryClientProvider` with context. If there is no context, the trpc react-query will throw `No QueryClient Set ....`. If there is only one `QueryClientProvider` with context, the existing react-query will throw `No QueryClient Set ....`.

The following code solves my issue.

trpc.ts
```javascript
export const context = createContext<QueryClient | undefined>(undefined);

export const trpc = createTRPCReact<AppRouter>({ reactQueryContext: context });
```

App.tsx
```diff
+ <trpc.Provider client={trpcClient} queryClient={queryClient}>
+    <QueryClientProvider client={queryClient} context={context}>
        <QueryClientProvider client={queryClient}>
            ...app
        </QueryClientProvider>
+    </QueryClientProvider>
+ </trpc.Provider>
```

Related:
[react-query v4 docs](https://tanstack.com/query/v4/docs/react/guides/migrating-to-react-query-4#custom-contexts-for-multiple-providers)
[another issue](https://github.com/trpc/trpc/issues/3316)

### 👨‍👧‍👦 Contributing

- [x] 🙋‍♂️ Yes, I'd be down to file a PR implementing the suggested changes!

**Comment by Nick-Lucas** [comment]
I believe it should be fine to re-use the existing react-query client in trpc. Just import it and use it with tRPC as well

**Comment by juliusmarminge** [comment, marked as answer]
Yea you definetely dont need multiple queryclients. As Nick Said you can reuse the existing one

**Comment by yikkwongwong** [comment]
I can reuse the queryclient but not the QueryClientProvider. If i pass it without context, trpc wrapper will throw `No QueryClient set`, but existing react-query can run just fine.

**Comment by Nick-Lucas** [comment]
~~Looking at your code example, I think you may need the react query provider above trpc in the hierarchy~~ actually that order looks fine, but you do need both providers to be at the top of your app hierarchy

**Reply by Nick-Lucas** [reply]
Can you share more of your app setup @yikkwongwong? It's likely a problem in how you're going about it

**Comment by juliusmarminge** [comment]
You shouldnt need the shared context either

**Comment by yikkwongwong** [comment]
Surprisingly the issue is most likely due to CRA. I migrated to Vite and everything works fine without using 2 QueryClientProvider and context.

**Reply by Nick-Lucas** [reply]
Shouldn't really be an issue as they don't change how react works at runtime in the browser,  but I'm glad that whatever tweaking you had to do also solved the problem :)

**Comment by sinclairnick** [comment]
@yikkwongwong I recently opened [an issue](https://github.com/trpc/trpc/issues/4012), following a discussion with Alex, about making the integration with RQ in general more ergonomic. Currently, IMO, TRPC encapsulates RQ functionality too much and should "sit beside" not "on top" of RQ. As such I've created a [trpc-react](https://github.com/sinclairnick/trpc-react) in the meantime to test out this DX.

It's more feature complete (for the features I'm interested in) and requires less adoption friction. But it also lacks docs, although intellisense should suffice. If/when it's developed/stable enough we may merge it into the trpc core repo.
